File: PythonSolutions/205.py
class Solution(object):
    def isIsomorphic(self, s, t):
        hashes = {}
        reservedChars = []
        for i in range(len(s)):
            if s[i] in hashes:
                if hashes[s[i]] != t[i]:
                    return False
            else:
                if t[i] in reservedChars:
                    return False
                hashes[s[i]] = t[i]
                reservedChars.append(t[i])
        return True


res = Solution()
print(res.isIsomorphic("badc", "baba"))

File: typescript-solutions/205.ts
function isIsomorphic(s: string, t: string): boolean {
  const characterMap: Record<string, string> = {};

  for (let i = 0; i < s.length; i++) {
    if (characterMap[s[i]]) {
      if (characterMap[s[i]] !== t[i]) return false;
    } else {
      if (Object.values(characterMap).some((val) => val === t[i])) return false;
      characterMap[s[i]] = t[i];
    }
  }

  return true;
}

console.log(isIsomorphic('badc', 'baba'));
